src/main/scala/com/microsoft/hyperspace/index/zordercovering/ZOrderField.scala

```diff
@@ -0,0 +1,569 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.zordercovering
+
+import java.util.BitSet
+
+import org.apache.spark.sql.catalyst.util
+import org.apache.spark.sql.types._
+
+import com.microsoft.hyperspace.HyperspaceException
+
+abstract class ZOrderField extends Serializable {
+  val name: String
+  val bitLen: Int
+  def getBitSet(valueAny: Any): BitSet
+  def throwOutOfRangeException(minVal: Any, maxVal: Any, value: Any): Unit = {
+    throw HyperspaceException(
+      "Unexpected operation on ZOrderField. Value out of range: " +
+        s"name=$name, value=$value, min=$minVal, max=$maxVal")
+  }
+
+  protected val emptyBitSet = new BitSet
+  protected def checkEmptySetEligible(valueAny: Any): Boolean = {
+    bitLen == 0 || valueAny == null
+  }
+}
+
+/**
+ * ZOrderField for numeric types based on percentile values of the original column values.
```

Comment:
While reading approach design blueprint in this comment, I am missing the critical component of thinking and design approach, covered indexes make sense in most situation, except, what is called, an lookup, when index is used to calculate row address to read additional data.
It would seem that in such a query case, particularly with large subquery passed in for matching, having unaligned partitions between index and outer rowstore, would result in shuffle, which would either reduce value of index used, or, depending on optimizer, prompt scan of rowstore, ignoring index altogether.
The rule inference for effective colocation of data was long thought out, and it seems like a very nice time to either tag on to this feature, or add it as a V2 for this feature, going forward.

Reply:
Are you suggesting partitioned covering index e.g. #509? Otherwise I don't really get the point of your comment..

Reply:
partitioning is not a "secret souse" in an of itself. By partitioning we don't get much win, unless partition is aligned ( colocated physically ) with outer row, the data row.
In fact this
org.apache.spark.sql.connector.catalog.InMemoryPartitionTable
pushes time to deliver this feature too close, to feel comfortable. I really want us to make it happen ASAP, rather than turn out late to finish line with someone already being past us.